keps/sig-architecture/4330-compatibility-versions/README.md

```diff
@@ -0,0 +1,1179 @@
+
+- (starting point) binary-version 1.28 (compat-version 1.28)
+- upgrade binary-version to 1.31 (compat-version stays at 1.28 - this is our skip-level binary upgrade)
+- keep binary-version 1.31 while upgrading compat-version to 1.29 (stepwise upgrade of compatibility)
+- keep binary-version 1.31 while upgrading compat-version to 1.30 (stepwise upgrade of compatibility)
+- keep binary-version 1.31 while upgrading compat-version to 1.31 (stepwise upgrade of compatibility)
+
+Benefits to upgrading binary version independent of compatibility version:
+
+- A skip-version binary upgrade that transitions through compat versions has
+  higher probability of bugs in those intermediate binary versions already being
+  fixed.
+- During an upgrade, it is possible upgrade the new binary version while the
+  compatibility version is fixed (e.g. `binaryVersion: 1.28` -> `{binaryVersion: 1.29, compatVersion: 1.28}`).
+  This allow differences in the binary (bugs fixed or introduced, performance
+  changes, ...) to be introduced into a cluster and verified safe before
+  allowing access to the new APIs and features new version. Once the binary
+  version has been deamed safe, the compatibility version can then be upgraded.
+- Any upgrade system that successfully detects failures between upgrade steps
+  can report which upgrade step failed. This makes it easier to diagnose the
+  failures, because there are fewer possible causes of the failure. (There's a
+  huge difference between "A cluster upgrade failed" and "A cluster upgrade
+  failed when only the apiserver's binary version changed").
+- For upgrades where multiple failures occur, this increases the odds those
+  failures are split across steps. An upgrade system that is able to pause after
+  a step where failures are detected allow for failures at that step to be
+  addressed before proceeding to subsequent steps. These failures can be
+  addressed without the disruption and "noise" from failures in subsequent
+  steps.
+- A compatibility version rollback can be performed without changing binary
+  version.
+
+
+### Goals
+
+- Introduce the metadata necessary to configure features/APIs/storage-versions/CEL
+  features to match the behavior of an older Kubernetes release version
+- A Kubernetes binary with compatibility version set to N, will pass the
+  conformance and e2e tests from Kubernetes release version N.
+- A Kubernetes binary with compatibility version set to N does not enable any
+  changes (storage versions, CEL feature, features) that would prevent it
+  from being rolled back to N-1.
+- The most recent Kubernetes version supports compatibility version being set to
+  the full range of supported versions (N..N-3).
+
+### Non-Goals
+
+- Support compatibility version for Alpha features.  Alpha feature are not
+  designed to be upgradable, so we will not allow alpha features to be used in
+  conjunction with compatibility version.
```

Comment:
Should I take this to mean that `--feature-gate=AlphaFeature` and `--compatibility-version=X` results in the apiserver refusing to run? Or that the alpha feature will be silently ignored/disabled? 

Is the expectation that people who use compatibility-version will only use it during their upgrade, and once their cluster is entirely done upgrading, remove that flag?

Reply:
The expectation is if `--feature-gate=AlphaFeature` and `--compatibility-version=X`, the alpha features for the current binary version would be enabled, regardless of which compatibility version.

Reply:
We want `--compatibility-version=X` to emulate the binary version of X with as much fidelity as possible. So we'd like `--feature-gate` to take the compatibility version into consideration when evaluating the flag value(s).  I.e. if `--feature-gate=AlphaFeature` would result an error on binary version X, it should result in the same error on other binary versions when `--compatibility-version=X`.

One nice property of doing it this way is that if someone has stepped their `--compatibility-version` up until it matches their binary version, they should be able to remove `--compatibility-version` with no risk of behavior changes.

Reply:
@jpbetz Is my understanding correct? 
if featureA is introduced as Alpha in N, and stayed alpha in N+1 but with some updates. I can still do `--feature-gate=featureA=true --compatibility-version=N`, but the user of this feature should anticipate the behavior of featureA to be the same as binaryVersion N+1 instead of the behavior before the update.

Reply:
What I'm proposing is that if the user provides `--feature-gate={any alpha feature}=true --compatibility-version={any valid compat version value}` that the apiserver exists with an error reporting that alpha features may not be used in conjunection with `--compatibility-version`.